src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/SymmetricAlgorithm.cs

```diff
@@ -428,6 +428,220 @@ public int GetCiphertextLengthCfb(int plaintextLength, PaddingMode paddingMode =
+        /// <param name="paddingMode">The padding mode used to produce the ciphertext and remove during decryption</param>
+        /// <param name="bytesWritten">When this method returns, the total number of bytes written to <paramref name="destination" /></param>
+        /// <exception cref="ArgumentOutOfRangeException">
+        ///   <paramref name="paddingMode" /> is not a valid padding mode.
+        /// </exception>
+        /// <exception cref="CryptographicException">
+        ///   <see cref="TryDecryptEcbCore" /> could not decrypt the ciphertext.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        ///   The buffer in <paramref name="destination"/> is too small to hold the plaintext data.
+        /// </exception>
+        /// <remarks>
+        ///   This method's behavior is defined by <see cref="TryDecryptEcbCore" />.
+        /// </remarks>
+        public bool TryDecryptEcb(ReadOnlySpan<byte> ciphertext, Span<byte> destination, PaddingMode paddingMode, out int bytesWritten)
+        {
+            CheckPaddingMode(paddingMode);
+            return TryDecryptEcbCore(ciphertext, destination, paddingMode, out bytesWritten);
+        }
+
+        public byte[] EncryptEcb(byte[] plaintext, PaddingMode paddingMode)
+        {
+            // paddingMode is validated by callee
+            if (plaintext is null)
+                throw new ArgumentNullException(nameof(plaintext));
+
+            return EncryptEcb(new ReadOnlySpan<byte>(plaintext), paddingMode);
+        }
+
+        public byte[] EncryptEcb(ReadOnlySpan<byte> plaintext, PaddingMode paddingMode)
+        {
+            CheckPaddingMode(paddingMode);
+
+            int ciphertextLength = GetCiphertextLengthEcb(plaintext.Length, paddingMode);
+
+            // We expect most if not all uses to encrypt to exactly the ciphertextLength
+            byte[] buffer = GC.AllocateUninitializedArray<byte>(ciphertextLength);
+
+            if (!TryEncryptEcbCore(plaintext, buffer, paddingMode, out int written))
+            {
+                // This means a user-derived imiplementation added more padding than we
+                // expected.
+                // TODO: better error message
+                throw new CryptographicException();
+            }
+
+            if (written != ciphertextLength)
+            {
+                // This should not happen, but could if a derived class's implementation of core
+                // did something non-standard (encrypt to a partial block). This can't happen for
+                // multiple padding blocks since the buffer would have been too small in the first
+                // place. It doesn't make sense to try and support partial block encryption, likely
+                // something went very wrong. So throw.
+
+                // TODO: better error message
+                throw new CryptographicException();
+            }
```

Comment:
Hm.  When I did non-cryptographic hashing I ran into this same problem.  The conclusion I came to there was to redefine the API to ignore it:  `TryEncryptEcbCore(plaintext, buffer, paddingMode, out int written)` => `EncryptEcbCore(plaintext, buffer.Slice(0, ciphertextLength), paddingMode)`, and it's up to the derived type to always write over every byte of the ciphertext span.

False isn't allowed.  Partial write isn't allowed.  So don't ask the questions.

Thoughts?

This is the only comment that I have that would be a reason to not merge today, so I'll post it special.  I have some nitpicks that I'll follow up with shortly, that I'd been intending for "let's get this into today's snapshot".

If you agree, having been in the bowels of the code here, there's a chance we can get the API update into API Review in ~2 hours, and maybe still get mode 1 merged in today's snapshot (unless it was yesterday's, then nevermind).

Reply:
@bartonjs

> there's a chance we can get the API update into API Review in ~2 hours

Ah crud. Sorry Tuesday was... busy... for me.

> it's up to the derived type to always write over every byte of the ciphertext span.

That would work for `Encrypt` I think (since we can calculate the ciphertext's length), but `Decrypt` might be a little weird because of padding removal. We don't know how long a plaintext is until we decrypt it. "write over every byte" to me means knowing the buffer length prior. I would like to keep the API symmetry between encrypt/decrypt.

There is a > 0% chance I am misunderstanding your suggestion though.

Reply:
Ah, right.  Depadding.  Which I already talked about.  Yep.

I agree it feels a little off to have `protected virtual int EncryptEcbCore(...)` and `protected virtual bool TryDecryptEcbCore(...)`, and that depadding belongs in the implementation types.  So... nevermind. :smile: